### DISCUSSION TITLE
How to set up an external OAuth provider, such as google or github?

### DISCUSSION THREAD

**Discussion by mwozniczak** [fastapi/fastapi#7711] [Questions] [answered]
### First check

* [✔] I used the GitHub search to find a similar issue and didn't find it.
* [✔] I searched the FastAPI documentation, with the integrated search.
* ✔] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

I wanted to migrate my app from flask to fastAPI.: One part I need to migrate is handling auth via an external provider (in my case, google). In flask, a library called [flask-dance](https://flask-dance.readthedocs.io/en/latest/) that makes it really easy to implement external auth that only requires me to provide the security-related credentials (which strikes me as a relatively common use case: I don't want to handle all the particulars of handling auth, but want to secure my app)

I've tried setting up external auth provider for my app using fastAPI. If I understand correctly, the documentation goes into details of a case where we provide the username and password, however in the case of google auth, this is handled on their side and my flask app didn't have anything to do with the users' passwords. I can't find any info on how to handle that use case.

**Comment by aniljaiswal** [comment]
I think you can find something useful here https://github.com/googleapis/google-api-python-client/blob/master/docs/auth.md

The `Auth` section has documentation on how to handle flows for all kinds of Oauth setups.

**Comment by aniljaiswal** [comment]
Additional information here if you have a web-based frontend and want to authenticate with your FastAPI based backend: https://developers.google.com/identity/sign-in/web/backend-auth

**Comment by jorgecarleitao** [comment]
Would the following help?

```
from starlette.applications import Starlette
from starlette_oauth2_api import AuthenticateMiddleware

app = Starlette()
app.add_middleware(AuthenticateMiddleware,
    {
        'example': {
            'uri': f'https://example.com/tenant-id/v2.0/.well-known/openid-configuration',
            'issuer': settings.EXAMPLE_ISSUER,
            'audience': settings.EXAMPLE_AUDIENCE,
        }
    },
)
```

using, https://pypi.org/project/starlette-oauth2-api/

**Comment by mwozniczak** [comment]
@jorgecarleitao it's definitely a start, yeah. but, being something of a noob to both starlette/fastapi and oauth, i don't entirely sure how this maps to what I know about setting up google-based auth. Like, what's `issuer`? what do i put in `audience`? How does this map to the client secret + secret ID I'm getting from google when I register the credentials provider?

@anil-grexit okay, but what you've linked to is somewhat low-level info on how the protocol works in general, but not how to set it up in the context of fastAPI. The reason I brought up flask-dance is that a) this is what I'm migrating from, and b) it's fairly trivial and clearly understandable how to set it up. I'm missing at least a write-up in fastAPI docs on how I would go about accomplishing what I'm currently doing with flask-dance and four-ish lines of code (most of which are pointing it to appropriate secret variables)

**Comment by jorgecarleitao** [comment]
@mwozniczak : it depends on what you are trying to achieve: `starlette-oauth2-api` assumes that the server (fastAPI) is an API that expects a `Authorization: Bearer <token>` from the client - i.e. the client fetched the access token from the identity provider e.g. on a single page application using implicit flow or PKCE. In this case, the server does not need to impersonate the user to use other APIs and therefore there is no need for a client secret.

AFAIK Google API has no support for access tokens to custom APIs - it only supports access tokens to APIs that google itself provides. The rational being that, as a developer on Google API, you are probably building a server that uses Google's API. To create access tokens for generic APIs on google AD, you would need to use [Google Cloud Identity](https://cloud.google.com/identity).

Regardless, you can still use Google API to give you an id token (i.e. who the user is) through openid, and manage authorization yourself. A working configuration for google would be:

```
        'google': {
                'uri': 'https://accounts.google.com/.well-known/openid-configuration',
                # see https://developers.google.com/identity/protocols/OpenIDConnect
                'issuer': 'https://accounts.google.com',
                # this was provisioned manually on Google console.
                'client_id': '<ID>.apps.googleusercontent.com',
                # google assumes that audience = client_id
                'audience': '<ID>.apps.googleusercontent.com',
                'scope': 'openid',
            },
```
where `<ID>` is a "OAuth 2.0 Client IDs" created on `GoogleAPI>Credentials`. At this point, the client can request an (id) token to google (via PKCE or implicit flow), and, once passed to server, it validates the token and puts the claims on the session (e.g. the user's email), which you can use on an authorization middleware.

The use-case with a client secret is used when the server (fastAPI) impersonates the user to do something on their behalf, by fetching (and storing) an access token pertaining to the user. That is useful when the API itself needs permissions to access another API (e.g. google compute) on the user's behalf. If this is your use-case, then [this client](https://docs.authlib.org/en/latest/client/starlette.html#using-fastapi) is probably what you need.

**Comment by mwozniczak** [comment, marked as answer]
@jorgecarleitao this might be it!

Your last link pointed me at [this example app](https://github.com/authlib/demo-oauth-client/tree/master/starlette-google-login) which _seems_ like it's doing _exactly_ what I was looking for! (but I will have to double-check, of course ;))

**Reply by abbazs** [reply]
How can I use it to protect the protected endpoints? How can I set a dependency to get the current user and validate if they are authorized?

**Comment by jorgecarleitao** [comment]
Great that it was of some help. @mwozniczak, If you think that the question is answered, feel free to mark it as closed so to keep the issue list up-to-date. :)

**Comment by mwozniczak** [comment]
@jorgecarleitao hey, apologies for the delay, life got in the way of my being able to answer ;)

about the only thing I can think of that miiight need doing here is maybe adding something to the documentation, because there might be others like me? but otherwise, yeah, let me close this and you guys do what you think is right ;)

**Comment by leemeador** [comment]
Does anyone have any ideas on how to test a service endpoint protected this way without requiring to generate a valid token and let the service make an external call to the provider (e.g. Google). Is there a way to mock out the middleware  in the test and test the service in a test that is not testing the security.

**Comment by jorgecarleitao** [comment]
@leemeador , check the tests [here](https://gitlab.com/jorgecarleitao/starlette-oauth2-api/-/blob/master/tests.py#L132): configure the middleware with a key, and pass that key to the header.
